### PR TITLE
Feat: add item search capability by its parameters & Feat: reuse categories and allergens by title

### PR DESCRIPTION
**Pull request by psukys**
Adds unconstrained search for item values by use request query parameters. Couldn't find any definitive guidelines or approach on REST search, thus implemented this way.

Simple use case would be (used for autocomplete):
`/items/?title=Sp`
which would find items with titles that contain Sp in it (i.e. Spaghetti)

Added reuse feature for categories and allergens. Also fixed `callback` usage: was used in cyclic logic, whereas async expects it to be used only once; now is called only at the end of `async.each`

Added Image field with it's handling.

cUrl request to check with (please modify image information accordingly):
```
curl -v -H "Content-Type:multipart/form-data" -F "image=@local_image.png;type=image/png" -F title="Lasagna" -F description="Some delicious food" -F categories='[{"title":"Italian cuisine","description":"made in Italy"}, {"title":"Meat","description":"This meal contains meat"}]' -F allergens='[{"title":"Random allergen", "description":"very allergic"}]' localhost:8000/items
```

### CODE REVIEW

File: src/controllers/ItemController.ts [hunk 1]
```diff
@@ -10,7 +10,15 @@ import * as mongoose from "mongoose";
 const async = require('async');
 
 export const getItems = (req: Request, res: Response) => {
-  Item.find({}, function (err: mongoose.Error, items: IItem[]) {
+  const searchParams = {};
+  if (Object.keys(req.query).length == 0) {
+    console.log("No key queries - simple retrieval");
+  } else {
+    Object.keys(req.query).forEach((key) => {
```
Comment (shumail92) [inline src/controllers/ItemController.ts]: what is format of query that is expected?

Comment (psukys) [inline src/controllers/ItemController.ts]: A simple query string, where  querystring's `field` represents mongo collection's `field` and querystring's `value` represents mongo collection document's value for a single field.

For example:
`/items/?title=Sp`

`title` is the `field` in `items` collection that will be looked upon for `Sp` value (match)

**Comment by psukys** [comment]
Added reference resolving for categories and allergens in item

**Comment by psukys** [comment]
Added image to model and logic behind posting image with new item

File: src/controllers/ItemController.ts [hunk 2]
```diff
@@ -40,53 +76,110 @@ export const postItem = (req: Request, res: Response) => {
...
-},
-      // Insert Item in collection
-    function (callback){
-  const item: IItem = new Item({
+    },
+    // save image
+    function (callback) {
+      if (req.body.image !== "") {
+        const image = decodeBase64Image(req.body.image);
+        namer("food", function (err, uniquePath) {
```
Comment (psukys) [inline src/controllers/ItemController.ts]: This gives an absolute path. While this is nice on localhost, it needs to have relative path or different approach on returning path

**Pull request merged**